packages/ui-components/src/components/PopupMenu/PopupMenu.component.tsx

```diff
@@ -0,0 +1,394 @@
+const sectionTitleStyles = `
+  jn-text-theme-light
+  jn-text-xs
+  jn-p-2
+  jn-cursor-default
+`
+
+const sectionSeparatorStyles = `
+  jn-h-
+  jn-bg-theme-background-lvl-3
+`
+
+// Make sure the floating reference wrapper element around the toggle fits snug around the toggle, so the positioning of the menu is correct relative to the visible toggle:
+const floatingReferenceWrapperStyles = `
+  jn-inline-flex
+`
+
+// ----- Interfaces -----
+
+// Type the PopupMenu context
+export interface PopupMenuContextType {
+  close: () => void
+  isOpen: boolean
+  menuSize: "normal" | "small"
+}
+
+export interface PopupMenuProps extends React.ComponentProps<typeof HeadlessMenu> {
+  /** The children to render. If no PopupMenu.Toggle child is passed, PopupMenu will render a default toggle. Pass a PopupMenu.Menu child in order to have a working component. */
+  children?: React.ReactNode
+  /** Add a custom className to the wrapping element. NOTE: The Menu will be rendered into a Portal outside the wrapping parent element, so the Menu and its children will be outside the scope of the parent CSS selector.*/
+  className?: string
+  /** Whether the PopupMenu is disabled. */
+  disabled?: boolean
+  /** The icon to render when using the default toggle. Will be ignored if a PopupMenu.Toggle child is passed. */
+  icon?: keyof typeof KnownIconsEnum
+  /** The size of the menu and its items. */
+  menuSize?: "normal" | "small"
+  /** Handler to run when the Menu closes. */
+  onClose?: () => void
+  /** Handler to run when the Menu opens. */
+  onOpen?: () => void
+}
+
+export interface PopupMenuToggleProps extends React.ComponentProps<typeof HeadlessMenu.Button> {
+  className?: string
+}
+
+//Extract types directly from React.ElementType as Headless Menu Items is a dynamically typed component that cannot be extended with React.ComponentPropsWithRef (such as `href` or `target` – these will be accepted when passing `as` as `a` in order to render an anchor, but not when rendering a div.):
+type HeadlessMenuItemsProps = React.ComponentPropsWithRef<React.ElementType>
+// Extend the extracted props instead:
+export interface PopupMenuMenuProps extends HeadlessMenuItemsProps {
+  as?: React.ElementType // Allow customizing the element type just as headless ui does
+  className?: string
+  children?: React.ReactNode
+  key?: React.Key
+}
+
+export interface PopupMenuItemProps extends React.ComponentProps<typeof HeadlessMenu.Item> {
+  as?: React.ElementType // Allow customising the element type to allow for rendering items as anchor elements, just as headless ui does
+  className?: string
+  disabled?: boolean
+  href?: string // accept a href when rendering a link, whether or not to pass it to HUI item will behandled in the component
+  icon?: keyof typeof KnownIconsEnum
+  label?: string
+  rel?: string // accept rel for anchors, handle in component
+  target?: string // accepot target for links, handle in component
+}
+
+export interface PopupMenuSectionProps {
```

Comment:
This component is quite big, I would suggest to extract all the types related code nearby in PopupMenu.types.ts file what do you think?

Reply:
Currently the standard is that types are defined in the same file as the component itself. I'm not against extracting them, but we should make a general decision and do it the same everywhere. We have always ensured to follow the same standard rules structurally with all components. We shouldn't break it in the context of this PR. Another one for #570